app/controllers/main/routes.py

```diff
@@ -37,38 +40,72 @@ def events(selectedTerm=None):
         currentTime = currentTime,
         user = g.current_user)
 
-@main_bp.route('/profile/<username>', methods = ['GET'])
-def profilePage(username):
-    if not g.current_user.isCeltsAdmin and not g.current_user.isCeltsStudentStaff and g.current_user.username != username:
-        return "Access Denied", 403
+@main_bp.route('/profile/<username>', methods=['GET'])
+def viewVolunteersProfile(username):
+    """
+    This function displays the information of a volunteer to the user
+    """
+    if (g.current_user.username == username) or g.current_user.isAdmin:
+         upcomingEvents = getUpcomingEventsForUser(username)
+         programs = Program.select()
+         interests = Interest.select().where(Interest.user == username)
+         programBan = ProgramBan.select().where(ProgramBan.user == username)
+         interests_ids = [interest.program for interest in interests]
+         eventParticipant = EventParticipant.select().where(EventParticipant.user == username)
+         currentTerm = Term.select().where(Term.isCurrentTerm == 1)
+         trainingEvents = (ProgramEvent.select().join(Event).where((Event.term == currentTerm) & (Event.isTraining == 1)))
+         trainingChecklist = {}
+         for program in programs:
+             trainingChecklist[program.id] = trainedParticipants(program.id)
+         eligibilityTable = []
+         for i in programs:
+             eligibilityTable.append({"program" : i,
+                                      "completedTraining" : (username in trainedParticipants(i)),
+                                      "isNotBanned" : isEligibleForProgram(i, username)})
+         return render_template ("/main/volunteerProfile.html",
+            programs = programs,
+            eventParticipant = eventParticipant,
+            interests = interests,
+            trainingEvents = trainingEvents,
+            programBan = programBan,
+            interests_ids = interests_ids,
+            trainingChecklist = trainingChecklist,
+            upcomingEvents = upcomingEvents,
+            eligibilityTable = eligibilityTable,
+            volunteer = User.get(User.username == username),
```

Comment:
You should probably get this up at the top and use it instead of username everywhere. That way if it is invalid you fail at the beginning, instead of at the end. It also makes the code clearer

Reply:
Refactored according to this suggestion

Reply:
Now you are doing this query twice. Once at the beginning and once here. And you still are using `username` everywhere instead of the user object.

Reply:
This comment was not resolved.